flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1194,32 +1194,34 @@ mlir::Value IntrinsicLibrary::genIAnd(mlir::Type resultType,
 }
 
 // ICHAR
-mlir::Value IntrinsicLibrary::genIchar(mlir::Type resultType,
-                                       llvm::ArrayRef<mlir::Value> args) {
+fir::ExtendedValue
+IntrinsicLibrary::genIchar(mlir::Type resultType,
+                           llvm::ArrayRef<fir::ExtendedValue> args) {
   // There can be an optional kind in second argument.
   assert(args.size() == 2);
+  auto charBox = args[0].getCharBox();
+  if (!charBox)
+    llvm::report_fatal_error("expected character scalar");
```

Comment:
The mlir emitError reports a nicer error message to the user. But report_fatal_error halts the compiler dead in its tracks. It seems like this type check should've already happened, so maybe a "crash immediately" is good enough here.

Reply:
I like that `mlir::emitError()` gives the user line programs, and not stopping is fine when we still manage to spit out an mlir module, but quite often we just go the undefined behavior after we emitted an error since we do not expect empty `mlir::Value` coming out from these functions. In debug mode, we usually hit an mlir/llvm assert quick soon after, but in release that can ends-up far and weird.

So my guideline is to use `mlir::emitError` when I know compilation will succeed but the compiled program will be wrong (let's say we are skipping a temp copy and we know we are wrong), and to use `llvm::report_fatal_error` if I cannot return a non empty `mlir::Value` from a function that should.

Is their a way to use mlir loc in `llvm::report_fatal_error` ?